samcli/lib/warnings/sam_cli_warning.py

```diff
@@ -83,3 +86,36 @@ def check(self, template_dict):
 
         send_warning = deployment_features_enabled_count > 0 and deployment_features_disabled_count > 0
         return (send_warning, self.WARNING_MESSAGE) if send_warning else (send_warning, "")
+
+
+class CodeDeployConditionWarning(TemplateWarning):
+    WARNING_MESSAGE = """Your template includes a deployment configuration with a Condition attached to it. SAM currently has a bug that
+ignores conditions for DeploymentPreference, causing CodeDeploy DeploymentGroups to be created in error.
+After October 23, 2020 the SAM service will fix this bug, causing subsequent deployments to remove these CodeDeploy DeploymentGroups
+if the attached Condition is false. For more information on this issue and how to mitigate it, please read these docs[1]
+
+[1] https://github.com/aws/aws-sam-cli/wiki/08-2020-codeploy-deploymentgroup-condition
+    """
+
+    def check(self, template_dict):
+        """
+        Checking if template dictionary have Function with Condition and DeploymentPreferences which
+        will trigger this warning.
+        """
+        functions = [
+            resource
+            for (_, resource) in template_dict.get("Resources", {}).items()
+            if resource.get("Type", "") == "AWS::Serverless::Function"
+        ]
```

Comment:
I think this should work, but we already have function provider which lists all the Functions from the template, should we use that? https://github.com/aws/aws-sam-cli/blob/39add9acd0c7532129a6543134919eb6ae9df4f3/samcli/lib/providers/sam_function_provider.py#L15